### PR TITLE
Hydra: Support for binding and mapping to gpu, pci and infiniband devices

### PR DESCRIPTION
**Pull request by KavithaTipturMadhu**

### CODE REVIEW

**Comment by KavithaTipturMadhu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by KavithaTipturMadhu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by KavithaTipturMadhu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by KavithaTipturMadhu** [comment]
@raffenet I have updated the PR with the changes we discussed. As for IO devices, currently the map and bind string have to match when an io device option is used as a binding directive. Mapping to io devices when binding is a non-io device is not allowed either.

**Comment by raffenet** [comment]
@KavithaTipturMadhu this looks to be working correctly when I bind to NICs like `ib0` and `eth0`. However, I still get an error when I try to bind to a pci id such as:
```
-bash-4.2$ ../../../../i/bin/mpiexec -bind-to pci:8086:24f0 ./a.out 
[proxy:0:0@skylake09.ftm.alcf.anl.gov] get_hw_obj_list (tools/topo/hwloc/topo_hwloc.c:200): unrecognized pci device id string "pci:8086:24f0"
[proxy:0:0@skylake09.ftm.alcf.anl.gov] handle_bitmap_binding (tools/topo/hwloc/topo_hwloc.c:346): error finding bind objects
[proxy:0:0@skylake09.ftm.alcf.anl.gov] HYDT_topo_hwloc_init (tools/topo/hwloc/topo_hwloc.c:550): error binding with bind "pci:8086:24f0" and map "(null)"
[proxy:0:0@skylake09.ftm.alcf.anl.gov] HYDT_topo_init (tools/topo/topo.c:62): unable to initialize hwloc
[proxy:0:0@skylake09.ftm.alcf.anl.gov] launch_procs (pm/pmiserv/pmip_cb.c:516): unable to initialize process topology
[proxy:0:0@skylake09.ftm.alcf.anl.gov] HYD_pmcd_pmip_control_cmd_cb (pm/pmiserv/pmip_cb.c:887): launch_procs returned error
[proxy:0:0@skylake09.ftm.alcf.anl.gov] HYDT_dmxu_poll_wait_for_event (tools/demux/demux_poll.c:77): callback returned error status
[proxy:0:0@skylake09.ftm.alcf.anl.gov] main (pm/pmiserv/pmip.c:200): demux engine error waiting for event
[mpiexec@skylake09.ftm.alcf.anl.gov] control_cb (pm/pmiserv/pmiserv_cb.c:208): assert (!closed) failed
[mpiexec@skylake09.ftm.alcf.anl.gov] HYDT_dmxu_poll_wait_for_event (tools/demux/demux_poll.c:77): callback returned error status
[mpiexec@skylake09.ftm.alcf.anl.gov] HYD_pmci_wait_for_completion (pm/pmiserv/pmiserv_pmci.c:194): error waiting for event
[mpiexec@skylake09.ftm.alcf.anl.gov] main (ui/mpich/mpiexec.c:338): process manager error waiting for completion
```
Is there something I'm doing wrong?

**Comment by KavithaTipturMadhu** [comment]
@raffenet, I looked at lstopo display from one of the skylake nodes and the pci device ids listed are 00:11.5, 00:17.0, 02:00.0, 3d:00.0, 3d:00.1, 5e:00.0, 86:00.0, af:00.0. I don't see the id 8086:24f0. Can you tell me how you you found out the pci device id? Also, can you try one of the ids I listed?

**Comment by raffenet** [comment]
Hmm, yes the ids you listed work fine. How do you get them? I don't see them when I do lstopo.

**Comment by KavithaTipturMadhu** [comment]
I listed the ids from lstopo-no-graphics utility I built using hwloc-2.0 sources, you can find it in /home/kmadhu/hwloc-build. The id you used is listed in /usr/bin/lstopo built from version 1.11.5rc2.

**Comment by raffenet** [comment]
Ah, I see. What you are using also matches what's given by `lspci`, so it must be right :smile:. I haven't given the code a thorough review yet, but it seems to be working as expected. I'll do the code next.

File: src/pm/hydra/tools/topo/hwloc/topo_hwloc.c [hunk 1]
```diff
@@ -172,55 +172,188 @@ static struct hwloc_obj_info_table hwloc_obj_info[] = {
     {NULL, HWLOC_OBJ_TYPE_MAX}
 };
 
-static HYD_status handle_bitmap_binding(const char *binding, const char *mapping)
+static HYD_status get_hw_obj_list(const char *mapping, hwloc_obj_t ** map_list,
```
Comment (raffenet) [inline src/pm/hydra/tools/topo/hwloc/topo_hwloc.c]: Are these argument names accurate anymore? Seems we are trying to get a list of objects from a string of format `object:count`. Not necessarily a mapping/map_list.

Comment (KavithaTipturMadhu) [reply src/pm/hydra/tools/topo/hwloc/topo_hwloc.c]: We get mapping object list as well as the number of objects and the hwloc object type.

Comment (raffenet) [reply src/pm/hydra/tools/topo/hwloc/topo_hwloc.c]: What I'm trying to say is that we pass both `binding` and `mapping` into the `mapping` argument. It seems like a slight mismatch in naming. Just a nitpick. Does that make sense?

Comment (KavithaTipturMadhu) [reply src/pm/hydra/tools/topo/hwloc/topo_hwloc.c]: ah yes, i'll rename the arguments

Comment (KavithaTipturMadhu) [reply src/pm/hydra/tools/topo/hwloc/topo_hwloc.c]: @raffenet I have updated the variable names.

Comment (raffenet) [reply src/pm/hydra/tools/topo/hwloc/topo_hwloc.c]: :+1:

**Comment by KavithaTipturMadhu** [comment]
@pavanbalaji Can you review this please?

**Comment by pavanbalaji** [comment]
@KavithaTipturMadhu there are a lot of warnings with `--enable-strict`:

```
  CC       tools/bootstrap/src/bsci_query_proxy_id.lo
../../../../mpich/src/pm/hydra/tools/topo/hwloc/topo_hwloc.c:205:25: warning: assigning to 'char *' from 'const char [4]' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
        obj_type_prefix = "pci";
                        ^ ~~~~~
../../../../mpich/src/pm/hydra/tools/topo/hwloc/topo_hwloc.c:211:25: warning: assigning to 'char *' from 'const char [4]' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
        obj_type_prefix = "gpu";
                        ^ ~~~~~
../../../../mpich/src/pm/hydra/tools/topo/hwloc/topo_hwloc.c:234:29: warning: assigning to 'char *' from 'const char [3]' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
            obj_type_prefix = "ib";
                            ^ ~~~~
```

**Comment by KavithaTipturMadhu** [comment]
@pavanbalaji I have updated the patches for all the warnings.

File: src/pm/hydra/ui/mpich/utils.c [hunk 2]
```diff
@@ -925,6 +925,16 @@ static void bind_to_help_fn(void)
...
+    printf
+        ("            pci:<id>         -- bind to processes on the non io ancestor domain of pci device with id\n");
+    printf
+        ("            gpu{<id>|:<n>}   -- bind to processes on the non io ancestor domain of gpu with index id or on non io ancestors of 'n' gpu devices\n");
+    printf
+        ("            ib{<id>|:<n>}    -- bind to processes on the non io ancestor domain of infiniband device with index id or on non io ancestors of 'n' ib devices\n");
+    printf
+        ("            en|eth{<id>|:<n>}    -- bind to processes on the non io ancestor domain of ethernet device with index id or on non io ancestors of 'n' ethernet devices\n");
+    printf
+        ("            hfi{<id>|:<n>}    -- bind to processes on the non io ancestor domain of omnipath device with index id or on non io ancestors of 'n' hfi devices\n");
```
Comment (pavanbalaji) [inline src/pm/hydra/ui/mpich/utils.c]: Please split this into two lines.  This looks very gross.

File: src/pm/hydra/tools/topo/hwloc/topo_hwloc.c [hunk 3]
```diff
@@ -189,22 +189,119 @@ static HYD_status get_hw_obj_list(const char *resource, hwloc_obj_t ** resource_
...
+                (*(resource + strlen("gpu")) == '\0' || *(resource + strlen("gpu")) == ':' ||
+                 atoi(resource + strlen("gpu"))
+                 == io_device->logical_index)) {
+                HYDU_REALLOC_OR_JUMP(resource_obj, hwloc_obj_t *,
+                                     (count + 1) * sizeof(hwloc_obj_t), status);
+                resource_obj[count++] = hwloc_get_non_io_ancestor_obj(topology, io_device);
+            }
+        }
+        *resource_list_length = count;
+
```
Comment (pavanbalaji) [inline src/pm/hydra/tools/topo/hwloc/topo_hwloc.c]: Unnecessary extra lines.

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by KavithaTipturMadhu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by KavithaTipturMadhu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by KavithaTipturMadhu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
@KavithaTipturMadhu: @raffenet might be busy with his PRs for the release.  I think you can remove his signoffs and merge this in.

File: src/pm/hydra/ui/mpich/utils.c [hunk 4]
```diff
@@ -954,6 +959,16 @@ static void bind_to_help_fn(void)
     printf("            l5cache{:<n>}    -- map by 'n' L5 cache domain\n");
     printf("            l5dcache{:<n>}   -- map by 'n' L5 data cache domain\n");
     printf("            l5ucache{:<n>}   -- map by 'n' L5 unified cache domain\n");
+    printf("            pci:<id>         -- map to non-io ancestor of PCI device\n");
```
Comment (raffenet) [inline src/pm/hydra/ui/mpich/utils.c]: should be "map by" to match the other options

**Pull request merged**